src/Microsoft.Android.Sdk.ILLink/TypeMappingStep.cs

```diff
@@ -35,55 +56,160 @@ protected override void ProcessAssembly (AssemblyDefinition assembly)
-			il.Emit (Mono.Cecil.Cil.OpCodes.Call, getTypeFromHandle);
-			il.Emit (Mono.Cecil.Cil.OpCodes.Callvirt, addMethod);
+			var il = method.Body.GetILProcessor ();
+
+			var orderedMapping = TypeMappings.Select(kvp => (Hash (kvp.Key), SelectTypeDefinition (kvp.Key, kvp.Value))).OrderBy (x => x.Item1);
+
+			var targets = new List<Instruction> ();
+			foreach (var name in names) {
+				targets.Add (il.Create (OpCodes.Ldstr, name));
+			}
+
+			il.Emit (OpCodes.Ldarg_0);
+			il.Emit (OpCodes.Switch, targets.ToArray ());
+
+			var defaultTarget = il.Create (OpCodes.Ldnull);
+			il.Emit (OpCodes.Br, defaultTarget);
+
+			foreach (var target in targets) {
+				il.Append (target);
+				il.Emit (OpCodes.Ret);
+			}
+
+			il.Append (defaultTarget);
+			il.Emit (OpCodes.Ret);
 		}
 
-		il.Emit (Mono.Cecil.Cil.OpCodes.Ret);
+		void GenerateHashes (ulong[] hashes)
+		{
+			// Sanity check: hashes must be unique and sorted
+			if (hashes.Length > 0) {
+				ulong previous = hashes[0];
+				for (int i = 1; i < hashes.Length; ++i) {
+					if (hashes[i] == previous) {
+						throw new InvalidOperationException ($"Duplicate hashes");
+					} else if (hashes[i] < previous) {
+						throw new InvalidOperationException ($"Hashes are not in ascending order");
+					}
+
+					previous = hashes[i];
+				}
+			}
+
+			// Create static array struct for `byte[#number-of-hashes]`
+			var arrayType = new TypeDefinition (
+				"",
+				"HashesArray",
+				TypeAttributes.NestedPrivate | TypeAttributes.ExplicitLayout,
+				module.ImportReference (typeof (ValueType)))
+			{
+				PackingSize = 1,
+				ClassSize = hashes.Length * sizeof (ulong),
+			};
+
+			type.NestedTypes.Add (arrayType);
+
+			// Create static field to store the raw bytes
+			var bytesField = new FieldDefinition ("s_hashes", FieldAttributes.Private | FieldAttributes.Static | FieldAttributes.InitOnly, arrayType);
+			bytesField.InitialValue = hashes.Select(h => BitConverter.GetBytes (h)).SelectMany (x => x).ToArray ();
+			if (!bytesField.Attributes.HasFlag (FieldAttributes.HasFieldRVA)) {
+				throw new InvalidOperationException ($"Field {bytesField.Name} does not have RVA");
+			}
+
+			type.Fields.Add (bytesField);
+
+			// Generate the Hashes getter
+			var getHashes = type.Methods.FirstOrDefault (f => f.Name == "get_Hashes") ?? throw new InvalidOperationException ($"Unable to find {TypeName}.get_Hashes field");
+
+			getHashes.Body.Instructions.Clear ();
+			var il = getHashes.Body.GetILProcessor ();
+
+			var genericUnsafeAsPointer = module.ImportReference (typeof (System.Runtime.CompilerServices.Unsafe).GetMethod("AsPointer"));
+			var unsafeAsPointer = new GenericInstanceMethod (genericUnsafeAsPointer);
+			unsafeAsPointer.GenericArguments.Add (module.ImportReference (typeof (byte)));
+
+			il.Emit (OpCodes.Ldsflda, bytesField);
+			il.Emit (OpCodes.Call, module.ImportReference (unsafeAsPointer));
```

Comment:
```suggestion
```
This is not necessary. It is valid IL to pass ref produced by Ldsflda into a method that takes `void*`. It is what Roslyn does for e.g. `ReadOnlySpan<byte> M() => [1, 2, 3, 4];`